### ISSUE TITLE
Future Releases

### ISSUE THREAD

**Issue by sebastianrothe** [mihar-22/svelte-jester#61]
ESM support in Jest is still very experimental https://jestjs.io/docs/ecmascript-modules

Some important things still don't work, like Mocking https://github.com/facebook/jest/issues/9430

For the new code to work, you need to enabled ESM support in Jest. Which means ALL steps described here https://jestjs.io/docs/ecmascript-modules

This setup can be quite a PITA, because you wil get problems with code written in CJS. This can be a burden to the user, that we enforce the switch to ESM for everyone using this library.

I think we need to refactor this code a little bit more and provide a timeline for different versions. My suggestion is:
- 1.x -> should still be CJS and support Jest 26
- 2.x -> should still be CJS, but support Jest 27 + asyncTransformers
- next -> switch to ESM

What do you think? @benmccann @mihar-22

**Comment by mihar-22** [comment]
Yep sounds good to me. I think generally we should always provide backwards support when upgrading to breaking changes such as what we did in 26 -> 27. Makes life easier, and we can deprecate stuff at a reasonable time in the future. I thought based on Jest docs it'll be fine but guess everything is far too experimental at this time.

**Comment by sebastianrothe** [comment]
At least, if there is Typescript in the mix as well.

**Comment by benmccann** [comment]
Sure, sticking with CJS sounds fine to me

**Comment by sebastianrothe** [comment]
I revised the versions once more:
-  1.x -> should still be CJS, synchronous and support Jest 26+27
-  2.x -> should still be CJS, but support Jest 27 + asyncTransformers only
-  next -> switch to ESM

**Comment by benmccann** [comment]
I don't know if it's worth creating more 1.x releases. The async transformer stuff should main be internal to this library and shouldn't cause difficulty for users upgrading, would it?

**Comment by sebastianrothe** [comment]
> I don't know if it's worth creating more 1.x releases. The async transformer stuff should main be internal to this library and shouldn't cause difficulty for users upgrading, would it?

No, you are right. If we implement it in CJS (or transpile) the user should not notice.

**Comment by sebastianrothe** [comment]
Well, it seems to be it more problematic, because of Sveltekit. Sveltekit uses adapters, which are ESM only (adapter-node, adapter-static). This means, if you have your svelte.config in CJS, then you are not able to use these adapters. Once you switch your config to ESM, *svelte-jester* will fail, if the config-loader is written in CJS.

I don't really know, how to proceed. Writing the whole library in ESM will force the usage of the flag. **--experimental-vm-modules**. Using CJS will fail for a svelte-config in ESM format.

We could provide multiple exports (one for ESM, one for CJS), but then we need to maintain two source folder. I can try to use the ESM version and then transpile it back to CJS, just for distribution.
